pkg/kubectl/cmd/util/openapi/openapi.go

```diff
@@ -0,0 +1,308 @@
+		definition := o.parseDefinition(name, d)
+		o.NameToDefinition[name] = definition
+		if len(definition.GroupVersionKind.Kind) > 0 {
+			o.GroupVersionKindToName[definition.GroupVersionKind] = name
+		}
+	}
+
+	if err := o.validate(); err != nil {
+		return nil, err
+	}
+
+	return o, nil
+}
+
+// validate makes sure the definition for each field type is found in the map
+func (o *Resources) validate() error {
+	types := sets.String{}
+	for _, d := range o.NameToDefinition {
+		for _, f := range d.Fields {
+			for _, t := range o.getTypeNames(f.TypeDefinition) {
+				types.Insert(t)
+			}
+		}
+	}
+	for _, n := range types.List() {
+		_, found := o.NameToDefinition[n]
+		if !found {
+			return fmt.Errorf("Unable to find definition for field of type %v", n)
+		}
+	}
+	return nil
+}
+
+func (o *Resources) getTypeNames(elem TypeDefinition) []string {
+	t := []string{}
+	if elem.IsKind {
+		t = append(t, elem.TypeName)
+	}
+	if elem.ElementType != nil && elem.ElementType.IsKind {
+		t = append(t, o.getTypeNames(*elem.ElementType)...)
+	}
+	return t
+}
+
+func (o *Resources) parseDefinition(name string, s spec.Schema) KindDefinition {
+	gvk, err := o.getGroupVersionKind(s)
+	value := KindDefinition{
+		Name:             name,
+		GroupVersionKind: gvk,
+		Extensions:       s.Extensions,
+		Fields:           map[string]FieldDefinition{},
+	}
+	if err != nil {
+		glog.Warning(err)
+	}
+
+	// Definition represents a primitive type - e.g.
+	// io.k8s.apimachinery.pkg.util.intstr.IntOrString
+	if o.isPrimitive(s) {
+		value.PrimitiveType = o.getTypeNameForField(s)
+	}
+	for fieldname, property := range s.Properties {
+		value.Fields[fieldname] = o.parseField(fieldname, property)
+	}
+	return value
+}
+
+func (o *Resources) buildElementFromSchema(s spec.Schema) TypeDefinition {
+	def := TypeDefinition{
+		TypeName:    o.getTypeNameForField(s),
+		IsPrimitive: o.isPrimitive(s),
+		IsArray:     o.isArray(s),
+		IsMap:       o.isMap(s),
+		IsKind:      o.isDefinitionReference(s),
+	}
+
+	if elementType, arrayErr := o.getElementType(s); arrayErr == nil {
+		d := o.buildElementFromSchema(elementType)
+		def.ElementType = &d
+	} else if valueType, mapErr := o.getValueType(s); mapErr == nil {
+		d := o.buildElementFromSchema(valueType)
+		def.ElementType = &d
+	}
+
+	return def
+}
+
+func (o *Resources) parseField(name string, s spec.Schema) FieldDefinition {
+	fieldDef := FieldDefinition{
+		Extensions:     s.Extensions,
+		TypeDefinition: o.buildElementFromSchema(s),
+	}
+	return fieldDef
+}
+
+// isArray returns true if s is an array type.
+func (o *Resources) isArray(s spec.Schema) bool {
+	return len(s.Type) > 0 && s.Type[0] == "array"
+}
+
+// isMap
+func (o *Resources) isMap(s spec.Schema) bool {
+	return len(s.Type) > 0 && s.Type[0] == "object"
+}
+
+// isPrimitive returns true if s is a primitive type
+func (o *Resources) isPrimitive(s spec.Schema) bool {
+	return !o.isArray(s) && !o.isMap(s) && !o.isDefinitionReference(s) && len(s.Type) > 0
```

Comment:
Updated the checks to be more explicit.

There is a unit check that test that every definition encountered has exactly 1 of (isPrimitive, isArray, isMap, isDefinitionReference) which should give some confidence that these are working as intended.